File: monolith/typescript/src/orders/orders.controller.ts
import { Body, Controller, Get, HttpCode, Param, Post, Res } from '@nestjs/common';
import type { Response } from 'express';
import { OrdersService, OrderItem } from './orders.service';

@Controller('v1/orders')
export class OrdersController {
  constructor(private readonly orders: OrdersService) {}

  @Post()
  create(@Body() body: { items: OrderItem[] }, @Res({ passthrough: true }) res: Response) {
    const order = this.orders.checkout(body?.items);
    res.status(201).setHeader('Location', `/v1/orders/${order.id}`);
    return order;
  }

  @Get(':id')
  get(@Param('id') id: string) {
    return this.orders.getOrder(id);
  }

  @Post(':id/cancel')
  @HttpCode(200)
  cancel(@Param('id') id: string) {
    return this.orders.cancel(id);
  }
}

File: monolith/java/src/main/java/io/backendguru/store/orders/OrdersController.java
package io.backendguru.store.orders;

import java.net.URI;
import java.util.List;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/v1/orders")
class OrdersController {

    private final OrdersService orders;

    OrdersController(OrdersService orders) {
        this.orders = orders;
    }

    record CreateOrderRequest(List<OrderItem> items) {
    }

    @PostMapping
    ResponseEntity<Order> create(@RequestBody CreateOrderRequest request) {
        Order order = orders.checkout(request == null ? null : request.items());
        return ResponseEntity.created(URI.create("/v1/orders/" + order.id())).body(order);
    }

    @GetMapping("/{id}")
    Order get(@PathVariable String id) {
        return orders.getOrder(id);
    }

    @PostMapping("/{id}/cancel")
    Order cancel(@PathVariable String id) {
        return orders.cancel(id);
    }
}
